{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from glob import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed files\n",
    "scv2_ref = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/scv2.fasta'\n",
    "gff = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/sc2_noCDS.gff'\n",
    "bed = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/nCoV-2019.primer.bed'\n",
    "adapters = '/shared/software/bbmap/v38.78/resources/adapters.fa'\n",
    "phiX = '/shared/software/bbmap/v38.78/resources/phix174_ill.ref.fa.gz'\n",
    "swift_masterfile = '~/bin/swift_sarscov2_v2_masterfile.txt'\n",
    "\n",
    "bb = '/shared/software/bin/bbduk.sh'\n",
    "ivar = 'singularity run docker://quay.io/biocontainers/ivar:1.3.1--h089eab3_0 ivar'\n",
    "primerclip = 'singularity exec docker://quay.io/biocontainers/primerclip:0.3.8--h9ee0642_1 primerclip'\n",
    "freebayes = 'singularity run https://depot.galaxyproject.org/singularity/freebayes:1.3.5--py39hba5d119_3 freebayes'\n",
    "picard = 'singularity run https://depot.galaxyproject.org/singularity/picard:2.25.4--hdfd78af_0 picard'\n",
    "#mutect2 = '/shared/software/bin/gatk Mutect2'\n",
    "mutect2 = 'singularity run https://depot.galaxyproject.org/singularity/gatk4:4.2.0.0--hdfd78af_1 gatk Mutect2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 3/20/21\n",
    "# rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_033021'\n",
    "# trimmedreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_033021/trimmed_reads'\n",
    "# mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_033021'\n",
    "\n",
    "# rawreads_sw = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_swift_033021'\n",
    "# trimmedreads_sw = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_swift_033021/trimmed_reads'\n",
    "# mapping_sw = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_swift_033021'\n",
    "\n",
    "\n",
    "# reads_artic_df = pd.read_csv('/Users/rosekantor/data/wbe_scv/qb3_march2021/artic_files.tsv', sep='\\t')\n",
    "# reads_swift_df = pd.read_csv('/Users/rosekantor/data/wbe_scv/qb3_march2021/swift_files.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 5/7/21\n",
    "# rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_050721'\n",
    "# trimmedreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_050721/trimmed_reads'\n",
    "# mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_050721'\n",
    "\n",
    "# outdir = '/Users/rosekantor/work/wbe_sequence_analysis/qb3_may2021'\n",
    "# # no sample names available yet, so starting processing with just a list of read names, added names2samples afterward\n",
    "# reads_artic_df = pd.read_csv('/Users/rosekantor/data/wbe_scv/qb3_may2021/reads_list.txt', names=['read1'])\n",
    "# names2samples = pd.read_csv('/Users/rosekantor/data/wbe_scv/qb3_may2021/names2samples.csv')\n",
    "\n",
    "# # create columns for read2 and sample_num\n",
    "# reads_artic_df['read2'] = reads_artic_df.read1.str.replace(r'_R1_', '_R2_')\n",
    "# reads_artic_df['Name'] = reads_artic_df.read1.str.split('_', expand=True)[0]\n",
    "\n",
    "# reads_artic_df = reads_artic_df.merge(names2samples, on='Name')\n",
    "# reads_artic_df['read2'] = reads_artic_df.read1.str.replace(r'_R1_', '_R2_')\n",
    "# reads_artic_df['sample'] = reads_artic_df.read1.str.split('_', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 5/26/21\n",
    "# rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_052621'\n",
    "# trimmedreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_052621/trimmed_reads'\n",
    "# mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_052621'\n",
    "# outdir = '/Users/rosekantor/work/wbe_sequence_analysis/qb3_artic_052621'\n",
    "# datadir = '/Users/rosekantor/data/wbe_scv/qb3_artic_052621/'\n",
    "\n",
    "# reads_artic_df = pd.read_csv('/Users/rosekantor/data/wbe_scv/qb3_artic_052621/reads_list.txt', names=['read1'])\n",
    "# reads_artic_df = reads_artic_df[~reads_artic_df.read1.str.contains('Undetermined')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 6/29/21\n",
    "# rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_062921'\n",
    "# trimmedreads = f'{rawreads}/trimmed_reads'\n",
    "# mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_062921'\n",
    "# outdir = '/Users/rosekantor/work/wbe_sequence_analysis/qb3_artic_062921'\n",
    "# datadir = '/Users/rosekantor/data/wbe_scv/qb3_artic_062921/'\n",
    "\n",
    "# reads_artic_df = pd.read_csv(f'{datadir}/reads_list.txt', names=['read1'])\n",
    "# reads_artic_df = reads_artic_df[~reads_artic_df.read1.str.contains('Undetermined')]\n",
    "\n",
    "# #create columns for read2 and sample_num\n",
    "\n",
    "# # for 6/29 need to keep the treatment in the name to distinguish them\n",
    "# reads_artic_df['read2'] = reads_artic_df.read1.str.replace(r'_R1_', '_R2_')\n",
    "# reads_artic_df['treatment'] = reads_artic_df.read1.str.split('_', expand=True)[0]\n",
    "# reads_artic_df['sample'] = reads_artic_df.read1.str.extract(r'(.+)_S\\d+_L001_R._001\\.fastq\\.gz')\n",
    "# reads_artic_df['sample_id'] = reads_artic_df.read1.str.extract(r'.+_(\\w+_\\w+_\\w+_\\d+_\\d)_.+')\n",
    "\n",
    "# # fill in the name of the sample we ran the isopropanol precipitation on\n",
    "# # (which was actually all 3 reps of K_K_INF from 5/30 and 5/31)\n",
    "# reads_artic_df.loc[reads_artic_df['sample'].str.contains('IPprep'), 'sample_id'] = 'K_K_INF_053021_1'\n",
    "\n",
    "# #reads_artic_df['sample'] = reads_artic_df.read1.str.extract(r'(\\w+_\\w+_\\w+_\\d+_\\d)_.+') # original\n",
    "# reads_artic_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7/30/21\n",
    "rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_073021'\n",
    "trimmedreads = f'{rawreads}/trimmed_reads'\n",
    "mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_073021'\n",
    "outdir = '/Users/rosekantor/work/wbe_sequence_analysis/qb3_artic_073021'\n",
    "datadir = '/Users/rosekantor/data/wbe_scv/qb3_artic_073021'\n",
    "\n",
    "reads_artic_df = pd.read_csv(f'{datadir}/reads_list.txt', names=['read1'])\n",
    "reads_artic_df = reads_artic_df[~reads_artic_df.read1.str.contains('Undetermined')]\n",
    "\n",
    "# create columns for read2 and sample_num\n",
    "\n",
    "# for 7/30, add the treatment in the name to distinguish them\n",
    "reads_artic_df['read2'] = reads_artic_df.read1.str.replace(r'_R1_', '_R2_')\n",
    "reads_artic_df['sample'] = reads_artic_df.read1.str.extract(r'(.+)_S\\d+_L001_R._001\\.fastq\\.gz')\n",
    "reads_artic_df['sample_id'] = reads_artic_df.read1.str.extract(r'(\\w+_\\w+_\\w+_\\d+_\\d).*')\n",
    "reads_artic_df['treatment'] = 'std'\n",
    "reads_artic_df.loc[reads_artic_df['sample'].str.contains('4A'), 'treatment'] = 'amicon'\n",
    "reads_artic_df.loc[reads_artic_df['sample'].str.contains('4N'), 'treatment'] = 'nanotrap'\n",
    "\n",
    "reads_artic_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read1</th>\n",
       "      <th>read2</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C_C_S68_071421_1_S13_L001_R1_001.fastq.gz</td>\n",
       "      <td>C_C_S68_071421_1_S13_L001_R2_001.fastq.gz</td>\n",
       "      <td>C_C_S68_071421_1</td>\n",
       "      <td>C_C_S68_071421_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C_C_S68_071421_2_S14_L001_R1_001.fastq.gz</td>\n",
       "      <td>C_C_S68_071421_2_S14_L001_R2_001.fastq.gz</td>\n",
       "      <td>C_C_S68_071421_2</td>\n",
       "      <td>C_C_S68_071421_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D_OSP_INF_071121_3_S11_L001_R1_001.fastq.gz</td>\n",
       "      <td>D_OSP_INF_071121_3_S11_L001_R2_001.fastq.gz</td>\n",
       "      <td>D_OSP_INF_071121_3</td>\n",
       "      <td>D_OSP_INF_071121_3</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D_OSP_INF_071421_1_S12_L001_R1_001.fastq.gz</td>\n",
       "      <td>D_OSP_INF_071421_1_S12_L001_R2_001.fastq.gz</td>\n",
       "      <td>D_OSP_INF_071421_1</td>\n",
       "      <td>D_OSP_INF_071421_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D_SEP_NF_071521_1_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>D_SEP_NF_071521_1_S3_L001_R2_001.fastq.gz</td>\n",
       "      <td>D_SEP_NF_071521_1</td>\n",
       "      <td>D_SEP_NF_071521_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>D_SEP_NF_071521_3_S4_L001_R1_001.fastq.gz</td>\n",
       "      <td>D_SEP_NF_071521_3_S4_L001_R2_001.fastq.gz</td>\n",
       "      <td>D_SEP_NF_071521_3</td>\n",
       "      <td>D_SEP_NF_071521_3</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G_G_INF_071321_1_S5_L001_R1_001.fastq.gz</td>\n",
       "      <td>G_G_INF_071321_1_S5_L001_R2_001.fastq.gz</td>\n",
       "      <td>G_G_INF_071321_1</td>\n",
       "      <td>G_G_INF_071321_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>G_G_INF_071321_2_S6_L001_R1_001.fastq.gz</td>\n",
       "      <td>G_G_INF_071321_2_S6_L001_R2_001.fastq.gz</td>\n",
       "      <td>G_G_INF_071321_2</td>\n",
       "      <td>G_G_INF_071321_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>G_G_MCS_071321_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>G_G_MCS_071321_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>G_G_MCS_071321_1</td>\n",
       "      <td>G_G_MCS_071321_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>G_G_MCS_071321_2_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>G_G_MCS_071321_2_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>G_G_MCS_071321_2</td>\n",
       "      <td>G_G_MCS_071321_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>J_J_UVANE_071521_1_S15_L001_R1_001.fastq.gz</td>\n",
       "      <td>J_J_UVANE_071521_1_S15_L001_R2_001.fastq.gz</td>\n",
       "      <td>J_J_UVANE_071521_1</td>\n",
       "      <td>J_J_UVANE_071521_1</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>J_J_UVANE_071621_2_S16_L001_R1_001.fastq.gz</td>\n",
       "      <td>J_J_UVANE_071621_2_S16_L001_R2_001.fastq.gz</td>\n",
       "      <td>J_J_UVANE_071621_2</td>\n",
       "      <td>J_J_UVANE_071621_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>K_K_INF_071121_2_S9_L001_R1_001.fastq.gz</td>\n",
       "      <td>K_K_INF_071121_2_S9_L001_R2_001.fastq.gz</td>\n",
       "      <td>K_K_INF_071121_2</td>\n",
       "      <td>K_K_INF_071121_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>K_K_INF_071521_2_S10_L001_R1_001.fastq.gz</td>\n",
       "      <td>K_K_INF_071521_2_S10_L001_R2_001.fastq.gz</td>\n",
       "      <td>K_K_INF_071521_2</td>\n",
       "      <td>K_K_INF_071521_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>L_L_INF_071221_3_S7_L001_R1_001.fastq.gz</td>\n",
       "      <td>L_L_INF_071221_3_S7_L001_R2_001.fastq.gz</td>\n",
       "      <td>L_L_INF_071221_3</td>\n",
       "      <td>L_L_INF_071221_3</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>L_L_INF_071421_2_S8_L001_R1_001.fastq.gz</td>\n",
       "      <td>L_L_INF_071421_2_S8_L001_R2_001.fastq.gz</td>\n",
       "      <td>L_L_INF_071421_2</td>\n",
       "      <td>L_L_INF_071421_2</td>\n",
       "      <td>std</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          read1  \\\n",
       "0     C_C_S68_071421_1_S13_L001_R1_001.fastq.gz   \n",
       "1     C_C_S68_071421_2_S14_L001_R1_001.fastq.gz   \n",
       "2   D_OSP_INF_071121_3_S11_L001_R1_001.fastq.gz   \n",
       "3   D_OSP_INF_071421_1_S12_L001_R1_001.fastq.gz   \n",
       "4     D_SEP_NF_071521_1_S3_L001_R1_001.fastq.gz   \n",
       "5     D_SEP_NF_071521_3_S4_L001_R1_001.fastq.gz   \n",
       "6      G_G_INF_071321_1_S5_L001_R1_001.fastq.gz   \n",
       "7      G_G_INF_071321_2_S6_L001_R1_001.fastq.gz   \n",
       "8      G_G_MCS_071321_1_S1_L001_R1_001.fastq.gz   \n",
       "9      G_G_MCS_071321_2_S2_L001_R1_001.fastq.gz   \n",
       "10  J_J_UVANE_071521_1_S15_L001_R1_001.fastq.gz   \n",
       "11  J_J_UVANE_071621_2_S16_L001_R1_001.fastq.gz   \n",
       "12     K_K_INF_071121_2_S9_L001_R1_001.fastq.gz   \n",
       "13    K_K_INF_071521_2_S10_L001_R1_001.fastq.gz   \n",
       "14     L_L_INF_071221_3_S7_L001_R1_001.fastq.gz   \n",
       "15     L_L_INF_071421_2_S8_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                          read2              sample  \\\n",
       "0     C_C_S68_071421_1_S13_L001_R2_001.fastq.gz    C_C_S68_071421_1   \n",
       "1     C_C_S68_071421_2_S14_L001_R2_001.fastq.gz    C_C_S68_071421_2   \n",
       "2   D_OSP_INF_071121_3_S11_L001_R2_001.fastq.gz  D_OSP_INF_071121_3   \n",
       "3   D_OSP_INF_071421_1_S12_L001_R2_001.fastq.gz  D_OSP_INF_071421_1   \n",
       "4     D_SEP_NF_071521_1_S3_L001_R2_001.fastq.gz   D_SEP_NF_071521_1   \n",
       "5     D_SEP_NF_071521_3_S4_L001_R2_001.fastq.gz   D_SEP_NF_071521_3   \n",
       "6      G_G_INF_071321_1_S5_L001_R2_001.fastq.gz    G_G_INF_071321_1   \n",
       "7      G_G_INF_071321_2_S6_L001_R2_001.fastq.gz    G_G_INF_071321_2   \n",
       "8      G_G_MCS_071321_1_S1_L001_R2_001.fastq.gz    G_G_MCS_071321_1   \n",
       "9      G_G_MCS_071321_2_S2_L001_R2_001.fastq.gz    G_G_MCS_071321_2   \n",
       "10  J_J_UVANE_071521_1_S15_L001_R2_001.fastq.gz  J_J_UVANE_071521_1   \n",
       "11  J_J_UVANE_071621_2_S16_L001_R2_001.fastq.gz  J_J_UVANE_071621_2   \n",
       "12     K_K_INF_071121_2_S9_L001_R2_001.fastq.gz    K_K_INF_071121_2   \n",
       "13    K_K_INF_071521_2_S10_L001_R2_001.fastq.gz    K_K_INF_071521_2   \n",
       "14     L_L_INF_071221_3_S7_L001_R2_001.fastq.gz    L_L_INF_071221_3   \n",
       "15     L_L_INF_071421_2_S8_L001_R2_001.fastq.gz    L_L_INF_071421_2   \n",
       "\n",
       "             sample_id treatment  \n",
       "0     C_C_S68_071421_1       std  \n",
       "1     C_C_S68_071421_2       std  \n",
       "2   D_OSP_INF_071121_3       std  \n",
       "3   D_OSP_INF_071421_1       std  \n",
       "4    D_SEP_NF_071521_1       std  \n",
       "5    D_SEP_NF_071521_3       std  \n",
       "6     G_G_INF_071321_1       std  \n",
       "7     G_G_INF_071321_2       std  \n",
       "8     G_G_MCS_071321_1       std  \n",
       "9     G_G_MCS_071321_2       std  \n",
       "10  J_J_UVANE_071521_1       std  \n",
       "11  J_J_UVANE_071621_2       std  \n",
       "12    K_K_INF_071121_2       std  \n",
       "13    K_K_INF_071521_2       std  \n",
       "14    L_L_INF_071221_3       std  \n",
       "15    L_L_INF_071421_2       std  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 8/7/21\n",
    "rawreads = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/raw/qb3_artic_080721'\n",
    "trimmedreads = f'{rawreads}/trimmed_reads'\n",
    "mapping_art = '/groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/qb3_artic_080721'\n",
    "outdir = '/Users/rosekantor/work/wbe_sequence_analysis/qb3_artic_080721'\n",
    "datadir = '/Users/rosekantor/data/wbe_scv/qb3_artic_080721'\n",
    "\n",
    "reads_artic_df = pd.read_csv(f'{datadir}/reads_list.txt', names=['read1'])\n",
    "reads_artic_df = reads_artic_df[~reads_artic_df.read1.str.contains('Undetermined')]\n",
    "\n",
    "# create columns for read2 and sample_num\n",
    "\n",
    "# for 7/30, add the treatment in the name to distinguish them\n",
    "reads_artic_df['read2'] = reads_artic_df.read1.str.replace(r'_R1_', '_R2_')\n",
    "reads_artic_df['sample'] = reads_artic_df.read1.str.extract(r'(.+)_S\\d+_L001_R._001\\.fastq\\.gz')\n",
    "reads_artic_df['sample_id'] = reads_artic_df.read1.str.extract(r'(\\w+_\\w+_\\w+_\\d+_\\d).*')\n",
    "reads_artic_df['treatment'] = 'std'\n",
    "\n",
    "reads_artic_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_artic_df.to_csv(f'{datadir}/samples_df.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline - ARTIC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make read trimming commands to remove adapters and low quality regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds = []\n",
    "for row in reads_artic_df.itertuples():\n",
    "    r1 = row.read1\n",
    "    r2 = row.read2\n",
    "    s = row.sample\n",
    "    trimR = f'{bb} -Xmx5g k=23 mink=11 hdist=1 tbo tpe t=2 ref={adapters} ktrim=r ftm=5 -in1={rawreads}/{r1} -in2={rawreads}/{r2} -out1={rawreads}/{s}.1.noadapt.fastq.gz -out2={rawreads}/{s}.2.noadapt.fastq.gz'\n",
    "    #filtphiX = f'{bb} -Xmx5g t=2 k=31 hdist=1 -in1={rawreads}/{s}.1.noadapt.fastq.gz -in2={rawreads}/{s}.2.noadapt.fastq.gz out1={rawreads}/{s}.1.nophix.fastq.gz out2={rawreads}/{s}.2.nophix.fastq.gz ref={phiX} stats={s}_stats_phix.txt'\n",
    "    qtrim = f'sickle pe -l 75 -f {rawreads}/{s}.1.noadapt.fastq.gz -r {rawreads}/{s}.2.noadapt.fastq.gz -t sanger -o {trimmedreads}/{s}.PE.1.fastq -p {trimmedreads}/{s}.PE.2.fastq -s {trimmedreads}/{s}.SR.fastq'\n",
    "    clean = f'rm {rawreads}/{s}.1.noadapt.fastq.gz {rawreads}/{s}.2.noadapt.fastq.gz {trimmedreads}/{s}.SR.fastq'\n",
    "    #qc = f'{fastqc} -t 2 {s}.PE.1.fastq {s}.PE.2.fastq'\n",
    "    gz1 = f'pigz -p 2 {trimmedreads}/{s}.PE.1.fastq'\n",
    "    gz2 = f'pigz -p 2 {trimmedreads}/{s}.PE.2.fastq'\n",
    "    cmd = [trimR, qtrim, clean, gz1, gz2]\n",
    "    cmd = \"; \".join(cmd)\n",
    "    cmds.append([s, cmd])\n",
    "\n",
    "prep_artic_df = pd.DataFrame(cmds, columns=[\n",
    "                        'sample',\n",
    "                        'read_trimming'\n",
    "                    ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make iVar pipeline for ARTIC data\n",
    "Note: not in mapping subdirectory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds = []\n",
    "for s in reads_artic_df['sample']:\n",
    "    mapping = f'bowtie2 -x {scv2_ref} -p 48 -1 {trimmedreads}/{s}.PE.1.fastq.gz -2 {trimmedreads}/{s}.PE.2.fastq.gz 2>{mapping_art}/{s}.log | sambam > {mapping_art}/{s}.bam'\n",
    "    map_qcmd = f'sbatch --wrap \"{mapping}\"'\n",
    "    \n",
    "    primer_trimming = f'{ivar} trim -e -q 0 -i {mapping_art}/{s}.bam  -b {bed} -p {mapping_art}/{s}.primer_trimmed 2>&1 | tee -a {mapping_art}/{s}.ivar_trim.log'\n",
    "    sort_index_bam = f'samtools sort {mapping_art}/{s}.primer_trimmed.bam > {mapping_art}/{s}.primer_trimmed.sorted.bam && samtools index {mapping_art}/{s}.primer_trimmed.sorted.bam'\n",
    "    samtools_depth = f'samtools depth {mapping_art}/{s}.primer_trimmed.sorted.bam > {mapping_art}/{s}.depth.txt'\n",
    "    mpileup_variants = f'samtools mpileup -aa -A  -B -Q 0 {mapping_art}/{s}.primer_trimmed.sorted.bam | {ivar} variants -p {mapping_art}/{s}.ivar -q 20 -t 0.03 -r {scv2_ref} -g {gff}'\n",
    "    freebayes_variants = f'{freebayes} -f {scv2_ref} -b {mapping_art}/{s}.primer_trimmed.sorted.bam -F 0.05 -C 2 --pooled-continuous > {mapping_art}/{s}.freebayes.vcf'\n",
    "    # gatk requires readgroup info to be added to the bam files, so we make this up\n",
    "    picard_readgrp = f'{picard} AddOrReplaceReadGroups I={mapping_art}/{s}.primer_trimmed.sorted.bam O={mapping_art}/{s}.primer_trimmed.sorted.rg.bam RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM={s}'\n",
    "    # index again... sigh\n",
    "    index_bam = f'samtools index {mapping_art}/{s}.primer_trimmed.sorted.rg.bam'\n",
    "    gatk_variants = f'{mutect2} --input {mapping_art}/{s}.primer_trimmed.sorted.rg.bam --reference {scv2_ref} --output {mapping_art}/{s}.gatk.vcf'\n",
    "    \n",
    "    cmds.append([s, map_qcmd, primer_trimming, sort_index_bam, samtools_depth,\n",
    "                 mpileup_variants, freebayes_variants, picard_readgrp, index_bam, gatk_variants])\n",
    "\n",
    "variant_df = pd.DataFrame(cmds, columns=[\n",
    "                        'sample',\n",
    "                        'mapping',\n",
    "                        'primer_trimming',\n",
    "                        'sort_index_bam',\n",
    "                        'samtools_depth',\n",
    "                        'mpileup_variants',\n",
    "                        'freebayes_variants',\n",
    "                        'picard_readgrp',\n",
    "                        'index_bam',\n",
    "                        'gatk_variants'\n",
    "                    ])\n",
    "cmds_artic_df = prep_artic_df.merge(variant_df, on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read_trimming</th>\n",
       "      <th>mapping</th>\n",
       "      <th>primer_trimming</th>\n",
       "      <th>sort_index_bam</th>\n",
       "      <th>samtools_depth</th>\n",
       "      <th>mpileup_variants</th>\n",
       "      <th>freebayes_variants</th>\n",
       "      <th>picard_readgrp</th>\n",
       "      <th>index_bam</th>\n",
       "      <th>gatk_variants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C_C_S68_071421_1</td>\n",
       "      <td>/shared/software/bin/bbduk.sh -Xmx5g k=23 mink...</td>\n",
       "      <td>bowtie2 -x /groups/banfield/projects/industria...</td>\n",
       "      <td>singularity run docker://quay.io/biocontainers...</td>\n",
       "      <td>samtools sort /groups/banfield/projects/indust...</td>\n",
       "      <td>samtools depth /groups/banfield/projects/indus...</td>\n",
       "      <td>samtools mpileup -aa -A  -B -Q 0 /groups/banfi...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>samtools index /groups/banfield/projects/indus...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C_C_S68_071421_2</td>\n",
       "      <td>/shared/software/bin/bbduk.sh -Xmx5g k=23 mink...</td>\n",
       "      <td>bowtie2 -x /groups/banfield/projects/industria...</td>\n",
       "      <td>singularity run docker://quay.io/biocontainers...</td>\n",
       "      <td>samtools sort /groups/banfield/projects/indust...</td>\n",
       "      <td>samtools depth /groups/banfield/projects/indus...</td>\n",
       "      <td>samtools mpileup -aa -A  -B -Q 0 /groups/banfi...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>samtools index /groups/banfield/projects/indus...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D_OSP_INF_071121_3</td>\n",
       "      <td>/shared/software/bin/bbduk.sh -Xmx5g k=23 mink...</td>\n",
       "      <td>bowtie2 -x /groups/banfield/projects/industria...</td>\n",
       "      <td>singularity run docker://quay.io/biocontainers...</td>\n",
       "      <td>samtools sort /groups/banfield/projects/indust...</td>\n",
       "      <td>samtools depth /groups/banfield/projects/indus...</td>\n",
       "      <td>samtools mpileup -aa -A  -B -Q 0 /groups/banfi...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>samtools index /groups/banfield/projects/indus...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D_OSP_INF_071421_1</td>\n",
       "      <td>/shared/software/bin/bbduk.sh -Xmx5g k=23 mink...</td>\n",
       "      <td>bowtie2 -x /groups/banfield/projects/industria...</td>\n",
       "      <td>singularity run docker://quay.io/biocontainers...</td>\n",
       "      <td>samtools sort /groups/banfield/projects/indust...</td>\n",
       "      <td>samtools depth /groups/banfield/projects/indus...</td>\n",
       "      <td>samtools mpileup -aa -A  -B -Q 0 /groups/banfi...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>samtools index /groups/banfield/projects/indus...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D_SEP_NF_071521_1</td>\n",
       "      <td>/shared/software/bin/bbduk.sh -Xmx5g k=23 mink...</td>\n",
       "      <td>bowtie2 -x /groups/banfield/projects/industria...</td>\n",
       "      <td>singularity run docker://quay.io/biocontainers...</td>\n",
       "      <td>samtools sort /groups/banfield/projects/indust...</td>\n",
       "      <td>samtools depth /groups/banfield/projects/indus...</td>\n",
       "      <td>samtools mpileup -aa -A  -B -Q 0 /groups/banfi...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "      <td>samtools index /groups/banfield/projects/indus...</td>\n",
       "      <td>singularity run https://depot.galaxyproject.or...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample                                      read_trimming  \\\n",
       "0    C_C_S68_071421_1  /shared/software/bin/bbduk.sh -Xmx5g k=23 mink...   \n",
       "1    C_C_S68_071421_2  /shared/software/bin/bbduk.sh -Xmx5g k=23 mink...   \n",
       "2  D_OSP_INF_071121_3  /shared/software/bin/bbduk.sh -Xmx5g k=23 mink...   \n",
       "3  D_OSP_INF_071421_1  /shared/software/bin/bbduk.sh -Xmx5g k=23 mink...   \n",
       "4   D_SEP_NF_071521_1  /shared/software/bin/bbduk.sh -Xmx5g k=23 mink...   \n",
       "\n",
       "                                             mapping  \\\n",
       "0  bowtie2 -x /groups/banfield/projects/industria...   \n",
       "1  bowtie2 -x /groups/banfield/projects/industria...   \n",
       "2  bowtie2 -x /groups/banfield/projects/industria...   \n",
       "3  bowtie2 -x /groups/banfield/projects/industria...   \n",
       "4  bowtie2 -x /groups/banfield/projects/industria...   \n",
       "\n",
       "                                     primer_trimming  \\\n",
       "0  singularity run docker://quay.io/biocontainers...   \n",
       "1  singularity run docker://quay.io/biocontainers...   \n",
       "2  singularity run docker://quay.io/biocontainers...   \n",
       "3  singularity run docker://quay.io/biocontainers...   \n",
       "4  singularity run docker://quay.io/biocontainers...   \n",
       "\n",
       "                                      sort_index_bam  \\\n",
       "0  samtools sort /groups/banfield/projects/indust...   \n",
       "1  samtools sort /groups/banfield/projects/indust...   \n",
       "2  samtools sort /groups/banfield/projects/indust...   \n",
       "3  samtools sort /groups/banfield/projects/indust...   \n",
       "4  samtools sort /groups/banfield/projects/indust...   \n",
       "\n",
       "                                      samtools_depth  \\\n",
       "0  samtools depth /groups/banfield/projects/indus...   \n",
       "1  samtools depth /groups/banfield/projects/indus...   \n",
       "2  samtools depth /groups/banfield/projects/indus...   \n",
       "3  samtools depth /groups/banfield/projects/indus...   \n",
       "4  samtools depth /groups/banfield/projects/indus...   \n",
       "\n",
       "                                    mpileup_variants  \\\n",
       "0  samtools mpileup -aa -A  -B -Q 0 /groups/banfi...   \n",
       "1  samtools mpileup -aa -A  -B -Q 0 /groups/banfi...   \n",
       "2  samtools mpileup -aa -A  -B -Q 0 /groups/banfi...   \n",
       "3  samtools mpileup -aa -A  -B -Q 0 /groups/banfi...   \n",
       "4  samtools mpileup -aa -A  -B -Q 0 /groups/banfi...   \n",
       "\n",
       "                                  freebayes_variants  \\\n",
       "0  singularity run https://depot.galaxyproject.or...   \n",
       "1  singularity run https://depot.galaxyproject.or...   \n",
       "2  singularity run https://depot.galaxyproject.or...   \n",
       "3  singularity run https://depot.galaxyproject.or...   \n",
       "4  singularity run https://depot.galaxyproject.or...   \n",
       "\n",
       "                                      picard_readgrp  \\\n",
       "0  singularity run https://depot.galaxyproject.or...   \n",
       "1  singularity run https://depot.galaxyproject.or...   \n",
       "2  singularity run https://depot.galaxyproject.or...   \n",
       "3  singularity run https://depot.galaxyproject.or...   \n",
       "4  singularity run https://depot.galaxyproject.or...   \n",
       "\n",
       "                                           index_bam  \\\n",
       "0  samtools index /groups/banfield/projects/indus...   \n",
       "1  samtools index /groups/banfield/projects/indus...   \n",
       "2  samtools index /groups/banfield/projects/indus...   \n",
       "3  samtools index /groups/banfield/projects/indus...   \n",
       "4  samtools index /groups/banfield/projects/indus...   \n",
       "\n",
       "                                       gatk_variants  \n",
       "0  singularity run https://depot.galaxyproject.or...  \n",
       "1  singularity run https://depot.galaxyproject.or...  \n",
       "2  singularity run https://depot.galaxyproject.or...  \n",
       "3  singularity run https://depot.galaxyproject.or...  \n",
       "4  singularity run https://depot.galaxyproject.or...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmds_artic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds_artic_df.read_trimming.to_csv(f'{outdir}/read_trimming_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.mapping.to_csv(f'{outdir}/mapping_artic.sh', index=False, header=False, quoting=csv.QUOTE_NONE)\n",
    "cmds_artic_df.primer_trimming.to_csv(f'{outdir}/primer_trimming_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.sort_index_bam.to_csv(f'{outdir}/sort_index_bam_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.samtools_depth.to_csv(f'{outdir}/samtools_depth_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.mpileup_variants.to_csv(f'{outdir}/mpileup_variants_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.freebayes_variants.to_csv(f'{outdir}/freebayes_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.picard_readgrp.to_csv(f'{outdir}/picard_readgrp_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.index_bam.to_csv(f'{outdir}/index_bam_artic.sh', index=False, header=False)\n",
    "cmds_artic_df.gatk_variants.to_csv(f'{outdir}/gatk_artic.sh', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gatk-launch CreateSequenceDictionary -R /groups/banfield/projects/industrial/wastewater_bayarea/2021/mapping/scv2.fasta'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# f'gatk CreateSequenceDictionary -R {scv2_ref}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline - Swift"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Swift"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds = []\n",
    "for row in reads_swift_df.itertuples():\n",
    "    r1 = row.read1\n",
    "    r2 = row.read2\n",
    "    s = row.sample\n",
    "    trimR = f'{bb} -Xmx5g k=23 mink=11 hdist=1 tbo tpe t=2 ref={adapters} ktrim=r ftm=5 -in1={rawreads_sw}/{r1} -in2={rawreads_sw}/{r2} -out1={rawreads_sw}/{s}.1.noadapt.fastq.gz -out2={rawreads_sw}/{s}.2.noadapt.fastq.gz'\n",
    "    #filtphiX = f'{bb} -Xmx5g t=2 k=31 hdist=1 -in1={rawreads}/{s}.1.noadapt.fastq.gz -in2={rawreads}/{s}.2.noadapt.fastq.gz out1={rawreads}/{s}.1.nophix.fastq.gz out2={rawreads}/{s}.2.nophix.fastq.gz ref={phiX} stats={s}_stats_phix.txt'\n",
    "    qtrim = f'sickle pe -l 75 -f {rawreads_sw}/{s}.1.noadapt.fastq.gz -r {rawreads_sw}/{s}.2.noadapt.fastq.gz -t sanger -o {trimmedreads_sw}/{s}.PE.1.fastq -p {trimmedreads_sw}/{s}.PE.2.fastq -s {trimmedreads_sw}/{s}.SR.fastq'\n",
    "    clean = f'rm {rawreads_sw}/{s}.1.noadapt.fastq.gz {rawreads_sw}/{s}.2.noadapt.fastq.gz {trimmedreads_sw}/{s}.SR.fastq'\n",
    "    #qc = f'{fastqc} -t 2 {s}.PE.1.fastq {s}.PE.2.fastq'\n",
    "    gz1 = f'pigz -p 2 {trimmedreads_sw}/{s}.PE.1.fastq'\n",
    "    gz2 = f'pigz -p 2 {trimmedreads_sw}/{s}.PE.2.fastq'\n",
    "    cmd = [trimR, qtrim, clean, gz1, gz2]\n",
    "    cmd = \"; \".join(cmd)\n",
    "    cmds.append([s, cmd])\n",
    "\n",
    "cmds_swift_df = pd.DataFrame(cmds, columns=[\n",
    "                        'sample',\n",
    "                        'read_trimming'\n",
    "                    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds_swift_df.read_trimming.to_csv('/Users/rosekantor/work/wbe_sequence_analysis/read_trimming_swift.sh', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds = []\n",
    "for s in reads_swift_df['sample']:\n",
    "    mapping = f'bowtie2 -x {scv2_ref} -p 16 --no-unal -1 {trimmedreads_sw}/{s}.PE.1.fastq.gz -2 {trimmedreads_sw}/{s}.PE.2.fastq.gz 2>{mapping_sw}/{s}.log | samtools sort -n -O sam -o {mapping_sw}/{s}.sam'\n",
    "    primer_trimming = f'{primerclip} {swift_masterfile} {mapping_sw}/{s}.sam {mapping_sw}/{s}.primer_trimmed.sam'\n",
    "    #sam2bam = \n",
    "    #sort_index_bam = f'samtools sort {s}.primer_trimmed.bam > {s}.primer_trimmed.sorted.bam && samtools index {s}.primer_trimmed.sorted.bam'\n",
    "    #mpileup_variants = f'samtools mpileup -aa -A  -B -Q 0 {s}.primer_trimmed.sorted.bam | singularity run docker://quay.io/biocontainers/ivar:1.3.1--h089eab3_0 ivar variants -p {s}.ivar -q 20 -t 0.03 -r scv2.fasta -g sc2_noCDS.gff'\n",
    "    cmds.append([s, mapping, primer_trimming])#([s, mapping, primer_trimming, sort_index_bam, mpileup_variants])\n",
    "\n",
    "variant_df = pd.DataFrame(cmds, columns=[\n",
    "                        'sample',\n",
    "                        'mapping',\n",
    "                        'primer_trimming',\n",
    "#                         'sort_index_bam',\n",
    "#                         'mpileup_variants'\n",
    "                    ])\n",
    "cmds_swift_df = cmds_swift_df.merge(variant_df, on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds_swift_df.read_trimming.to_csv('/Users/rosekantor/work/wbe_sequence_analysis/read_trimming_swift.sh', index=False, header=False)\n",
    "cmds_swift_df.mapping.to_csv('/Users/rosekantor/work/wbe_sequence_analysis/mapping_swift.sh', index=False, header=False)\n",
    "cmds_swift_df.primer_trimming.to_csv('/Users/rosekantor/work/wbe_sequence_analysis/primer_trimming_swift.sh', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
